README.md
# motorcortex-comets

## Demo
[Check it out here](https://kissmybutton.github.io/motorcortex-comets/demo/index.html)

## Installation

```bash
$ npm install --save @kissmybutton/motorcortex-comets
# OR
$ yarn add @kissmybutton/motorcortex-comets
```

## Loading

```javascript
const MotorCortex = require("@kissmybutton/motorcortex/");
const cometsDefinition = require("@kissmybutton/motorcortex-comets");
const Plugin = MotorCortex.loadPlugin(BannersDefinition);
```

# Create incident

## Comets

```javascript
const comet = new MyPlugin.Comets({
    width: 800,
    height: 300,
    cometMaxSize:581,
    cometMinSize:200,
    meteoriteColors:["#FFFFFF","#FFDC63"],
    auraIner:["#57DDBA","#62D8B9","#7CC8B3"],
    auraOuter:["#47F7C5","#87B3C1"],
    items: 3,
    repeats:5,
    duration:8000
}, {
    selector: '.comet1',
});
```

### Comets Attrs

| Name        | Are           | Values  |
| ------------- |:-------------:| -----:|
| width     | width of incident  | num |
| height     | height of incident  | num |
| cometMaxSize   | the maximum size of comets  | num |
| cometMinSize     | the minimum size of comets  | num |
| meteoriteColors     | The list of colors of meteorite  | array of colors |
| auraIner  | The list of colors for the iner aura | array of colors |
| auraOuter  | The list of colors for the iner aura | array of colors |
| items     | how meny commet will fall  | num |
| repeats     | how meny times clip wiil loop  | num |
|duration  | max duration of one comet fall | num |

## Stars

```javascript
const stars = new MyPlugin.Stars({
    width: 800,
    height: 300,
    starMaxSize:60,
    starMinSize:15, 
    starColor: "#F0F0F0",
    starGlowColor:"0 0 0 0 0.950946003   0 0 0 0 0.81265567   0 0 0 0 0.51528336  0 0 0 1 0",
    items: 15,
    repeats:5,
    duration:8000
}, {
    selector: '.comet2',
});
```

### Stars Attrs

| Name        | Are           | Values  |
| ------------- |:-------------:| -----:|
| width     | width of incident  | num |
| height     | height of incident  | num |
| starMaxSize   | the maximum size of each start  | num |
| starMinSize     | the minimum size of each start  | num |
| starColor     | the color of a star  | color |
| starGlowColor  | the color of star clow | sting of ColorMatrix value |
| items     | start number  | num |
| repeats     | how meny times clip wiil loop  | num |
|duration  | max duration of one full blinking animation | num |


CHANGELOG.md
# Changelog

### 1.0.4 (2021-05-12)


### Miscellaneous Chores

* release 1.0.4 ([20d737e](https://www.github.com/kissmybutton/motorcortex-comets/commit/20d737e3f1f006e86916ee3b827d1290b6adb6cb))


src/Incidents/Stars.js
import { HTMLClip, loadPlugin, Combo } from "@kissmybutton/motorcortex";
import AnimeDefinition from "@kissmybutton/motorcortex-anime";
const Anime = loadPlugin(AnimeDefinition);
import { spatial, timely } from "../helpers/randomizer";

export default class Stars extends HTMLClip {
  get html() {
    const attrs = spatial({
      dimensions: [
        [0, this.attrs.width],
        [0, this.attrs.height],
        [this.attrs.starMinSize, this.attrs.starMaxSize],
      ],
      divisions: [
        Math.floor(this.attrs.items / 2),
        Math.floor(this.attrs.items / 2),
        this.attrs.items,
      ],
      numberOfElements: this.attrs.items,
    });

    let list = [];
    for (let i = 0; i < this.attrs.items; i++) {
      const size = attrs[i][2];
      const top = attrs[i][1];
      const left = attrs[i][0];
      const comet = ` 
              <svg class="stars-svg stars-svg-${i}" width="${size}px" height="${size}px" style="left: ${left}px; top:${top}px;" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 50 50.15">
                <defs>
                  <radialGradient id="radial-gradient" cx="25.98" cy="26.84" r="22.27" gradientUnits="userSpaceOnUse">
                    <stop offset="0" stop-color="#fff9a6" stop-opacity="0.9"/>
                    <stop offset="0.07" stop-color="#fffab3" stop-opacity="0.77"/>
                    <stop offset="0.2" stop-color="#fffbc7" stop-opacity="0.57"/>
                    <stop offset="0.34" stop-color="#fffcd8" stop-opacity="0.4"/>
                    <stop offset="0.47" stop-color="#fffde6" stop-opacity="0.25"/>
                    <stop offset="0.6" stop-color="#fffef1" stop-opacity="0.14"/>
                    <stop offset="0.74" stop-color="#fffff9" stop-opacity="0.06"/>
                    <stop offset="0.87" stop-color="#fffffd" stop-opacity="0.02"/>
                    <stop offset="1" stop-color="#fff" stop-opacity="0"/>
                  </radialGradient>
                </defs>
                <title>Asset 10</title>
                <g id="Layer_2" data-name="Layer 2">
                  <g id="Layer_2-2" data-name="Layer 2">
                    <g>
                      <circle cx="25.98" cy="26.84" r="22.27" fill="url(#radial-gradient)"/>
                      <path id="stars-svg-path-1" d="M.15,26.27q13.13-2,18.73-4.52Q24.31,19.32,26,.16h0A.18.18,0,0,1,26.14,0a.19.19,0,0,1,.17.17q.6,17.88,6.81,21.58t16.72,4.62h0a.18.18,0,0,1,.16.2.19.19,0,0,1-.16.16Q38.1,27.86,33.12,32.05T26.49,50h0a.18.18,0,0,1-.36,0Q25.31,36.65,18.88,32T.17,26.62h0A.17.17,0,0,1,0,26.43.18.18,0,0,1,.15,26.27Z" fill="${this.attrs.starColor}" fill-rule="evenodd"/>
                    </g>
                  </g>
                </g>
              </svg>`;
      list += comet;
    }
    return `
    <div class="wrapper">
      ${list}
    </div>
    
    `;
  }

  get css() {
    return `
    .wrapper{
      width: ${this.attrs.width}px;
      height:${this.attrs.height}px;
      
    }
    .stars-svg{
      position: absolute;
      z-index: 3;

    }

  `;
  }

  buildTree() {
    function _getAnimatedAttr(value, opacity = false) {
      if (opacity === true) {
        return {
          opacity: value,
        };
      } else {
        return {
          transform: {
            scale: value,
          },
        };
      }
    }

    const timeData = timely({
      duration: this.attrs.duration / this.attrs.repeats,
      divisions: 3,
      numberOfElements: this.attrs.items,
      minDuration: 0.2,
      maxDuration: 1,
    });

    for (let i = 0; i < this.attrs.items; i++) {
      const starScale = new Combo(
        {
          incidents: [
            {
              incidentClass: Anime.Anime,
              attrs: {
                animatedAttrs: _getAnimatedAttr(
                  0,
                  this.attrs.blinkType === "opacity"
                ),
              },
              props: {
                duration: Math.round(timeData[i][1] / 2),
              },
              position: 0,
            },
            {
              incidentClass: Anime.Anime,
              attrs: {
                animatedAttrs: _getAnimatedAttr(
                  1,
                  this.attrs.blinkType === "opacity"
                ),
              },
              props: {
                duration: Math.round(timeData[i][1] / 2),
              },
              position: Math.round(timeData[i][1] / 2),
            },
          ],
        },
        {
          selector: `.stars-svg-${i}`,
          delay: timeData[i][0],
          repeats: this.attrs.repeats,
        }
      );
      this.addIncident(starScale, 0);
    }
  }
}


src/Incidents/Comets.js
import { HTMLClip, loadPlugin } from "@kissmybutton/motorcortex";
import AnimeDefinition from "@kissmybutton/motorcortex-anime";
const Anime = loadPlugin(AnimeDefinition);
import { _addDimension, timely } from "../helpers/randomizer";
const closeAngle = 29;

export default class Comets extends HTMLClip {
  get html() {
    // first get the time (position / duration) random numbers
    this.comets = timely({
      duration: this.attrs.duration / this.attrs.repeats,
      numberOfElements: this.attrs.items,
      minDuration: 0.1,
      maxDuration: 0.9,
    });
    // console.log(JSON.parse(JSON.stringify(this.comets)));
    // secondly depending on the duration set size
    for (let i = 0; i < this.comets.length; i++) {
      const comet = this.comets[i];
      const fraction = comet[0] / (this.attrs.duration / this.attrs.repeats);
      const size =
        (this.attrs.cometMaxSize - this.attrs.cometMinSize) * (1 - fraction) +
        this.attrs.cometMinSize;
      comet.push(size); // now each row has the form [duration, position, size]
    }
    // finally we add the position random dimension. Comets can start from anyplace that will allow them to be
    // visible in our Clip even a little and even in only part of their path
    const angle = ((90 - closeAngle) * Math.PI) / 180;
    const extraWidth = this.attrs.height * Math.tan(angle);
    this.comets = _addDimension(
      {
        from: 0,
        to: this.attrs.width + extraWidth,
        numberOfElements: this.attrs.items,
      },
      this.comets
    ); // now each row has the form [duration, position, size, left]
    // console.log(this.comets);

    let list = [];
    this.itemData = [];
    const A = ((90 - closeAngle) * Math.PI) / 180; // the big angle
    const B = (closeAngle * Math.PI) / 180; // the close angle
    for (let i = 0; i < this.attrs.items; i++) {
      const size = this.comets[i][2];
      const a = size * Math.sin(B); // the horizontal size of the comet box
      const b = size * Math.sin(A); // the vertical size of the comet box
      const left = this.comets[i][3];
      this.itemData.push({ left, top: -b, width: a, size });
      const comet = ` 
                <svg class="comet-svg comet-svg-${i}" style="left: ${left}px; top: -${b}px; width: ${a}px; height: ${b}px;" xmlns="http://www.w3.org/2000/svg" class="comet-green-svg" data-name="Layer 1" viewBox="0 0 450 270.44">
                <defs>
                    <linearGradient id="b" x1="-56.99" x2="-56.74" y1="394.68" y2="394.93" gradientTransform="matrix(1363.47 0 0 -819.42 77776.05 323638.12)" gradientUnits="userSpaceOnUse">
                    <stop offset="0" stop-color="${this.attrs.auraOuter[0]}" stop-opacity=".59"/>
                    <stop offset="1" stop-color="${this.attrs.auraOuter[1]}" stop-opacity="0"/>
                    </linearGradient>
                    <linearGradient id="d" x1="-56.97" x2="-56.79" y1="394.95" y2="395.11" gradientTransform="matrix(900.62 0 0 -542.55 51377.26 214502.16)" gradientUnits="userSpaceOnUse">
                    <stop offset="0" stop-color="${this.attrs.auraIner[0]}"/>
                    <stop offset=".31" stop-color="${this.attrs.auraIner[1]}" stop-opacity=".97"/>
                    <stop offset="1" stop-color="${this.attrs.auraIner[2]}" stop-opacity="0"/>
                    </linearGradient>
                    <linearGradient id="f" x1="-56.81" x2="-56.49" y1="399.49" y2="399.76" gradientTransform="matrix(116.79 0 0 -75.59 6649.18 30453.64)" gradientUnits="userSpaceOnUse">
                    <stop offset="0" stop-color="${this.attrs.meteoriteColors[0]}"/>
                    <stop offset="1" stop-color="${this.attrs.meteoriteColors[1]}"/>
                    </linearGradient>
                    <mask id="a" width="450" height="270.44" x="0" y="0" maskUnits="userSpaceOnUse">
                    <path fill="#fff" fill-rule="evenodd" d="M11.33 230a21.52 21.52 0 0020.31 38L450 38 429.7 0 11.33 230z"/>
                    </mask>
                    <mask id="c" width="297.24" height="179.06" x="7.51" y="84.83" maskUnits="userSpaceOnUse">
                    <path fill="#fff" fill-rule="evenodd" d="M14.65 238.41a13.55 13.55 0 0012.78 23.89l277.32-153.58L292 84.83 14.65 238.41z"/>
                    </mask>
                    <mask id="e" width="38.55" height="24.95" x="12.81" y="234.29" maskUnits="userSpaceOnUse">
                    <path fill="#fff" fill-rule="evenodd" d="M17.69 241.85a9.24 9.24 0 008.73 16.29c4.5-2.41 24.94-23.85 24.94-23.85s-29.17 5.15-33.67 7.56"/>
                    </mask>
                </defs>
                <g mask="url(#a)">
                    <path fill="url(#b)" fill-rule="evenodd" d="M11.33 230a21.52 21.52 0 0020.31 38L450 38 429.7 0 11.33 230z"/>
                </g>
                <g mask="url(#c)">
                    <path fill="url(#d)" fill-rule="evenodd" d="M14.65 238.41a13.55 13.55 0 0012.78 23.89l277.32-153.58L292 84.83 14.65 238.41z"/>
                </g>
                <g mask="url(#e)">
                    <path fill="url(#f)" fill-rule="evenodd" d="M17.69 241.85a9.24 9.24 0 008.73 16.29c4.5-2.41 24.94-23.85 24.94-23.85s-29.17 5.15-33.67 7.56"/>
                </g>
                </svg>
            `;

      list += comet;
    }

    return `
            <div class="wrapper">
            ${list}
            </div>
        `;
  }

  get css() {
    return `
            .wrapper{
                width: ${this.attrs.width}px;
                height:${this.attrs.height}px;
            }
            .comet-svg{
                position: absolute;
                z-index: 3;
            }
        `;
  }

  buildTree() {
    for (let i = 0; i < this.comets.length; i++) {
      // console.log(Math.tan(angle)*(-this.itemData[i].width -this.itemData[i].left)+ this.itemData[i].top)
      const angle = (-closeAngle * Math.PI) / 180;
      const moveMagentComet = new Anime.Anime(
        {
          animatedAttrs: {
            left: `-${this.itemData[i].width}px`,
            top: `${
              Math.tan(angle) *
                (-this.itemData[i].width - this.itemData[i].left) +
              this.itemData[i].top
            }px`,
          },
        },
        {
          selector: `.comet-svg-${i}`,
          duration: this.comets[i][0],
          delay: this.comets[i][1],
          repeats: this.attrs.repeats,
        }
      );
      this.addIncident(moveMagentComet, 0);
    }
  }
}


src/helpers/randomizer.js
const defaultDivisions = 4;
const maxDurationDefault = 0.6;
const minDurationDefault = 0.2;

function _shuffle(array) {
  let currentIndex = array.length,
    temporaryValue,
    randomIndex;

  // While there remain elements to shuffle...
  while (0 !== currentIndex) {
    // Pick a remaining element...
    randomIndex = Math.floor(Math.random() * currentIndex);
    currentIndex -= 1;

    // And swap it with the current element.
    temporaryValue = array[currentIndex];
    array[currentIndex] = array[randomIndex];
    array[randomIndex] = temporaryValue;
  }

  return array;
}

/*
params: {
    from,
    to,
    numberOfElements,
    divisions
}
*/
export function _addDimension(params, points) {
  if (points === null) {
    points = [];
    for (let i = 0; i < params.numberOfElements; i++) {
      points.push([]);
    }
  }
  const divisions = params.divisions || defaultDivisions;
  const lastRowElements = params.numberOfElements % divisions;
  const fullRows = (params.numberOfElements - lastRowElements) / divisions;
  const step = (params.to - params.from) / divisions;

  const divisionsByElement = [];
  for (let row = 0; row < fullRows; row++) {
    for (let i = 0; i < divisions; i++) {
      // for each division
      const limits = {
        from: params.from + step * i,
        to: params.from + step * (i + 1),
      };
      divisionsByElement.push(limits);
    }
  }

  if (lastRowElements !== 0) {
    const lastRowStep = (params.to - params.from) / lastRowElements;
    for (let i = 0; i < lastRowElements; i++) {
      const limits = {
        from: params.from + lastRowStep * i,
        to: params.from + lastRowStep * (i + 1),
      };
      divisionsByElement.push(limits);
    }
  }

  for (let i = 0; i < divisionsByElement.length; i++) {
    points[i].push(
      Math.random() * (divisionsByElement[i].to - divisionsByElement[i].from) +
        divisionsByElement[i].from
    );
  }

  return _shuffle(points);
}

/*
params: {
    dimensions: [[from, to], [from, to], ...],
    
    //the divisions per dimension. Each dimension is going to produce random 
    positions but all of these positions are going to equally be splitted into 
    a specific number of divisions. //
    
    divisions: [<integer>, <integer>, ...], 
    nunmberOfElements: <integer>  
}
*/
export function spatial(params) {
  if (!params.divisions) {
    params.divisions = [];
  }
  let points = null;
  for (let i = 0; i < params.dimensions.length; i++) {
    const limits = params.dimensions[i];
    const divisions = params.divisions[i] || defaultDivisions;
    points = _addDimension(
      {
        from: limits[0],
        to: limits[1],
        numberOfElements: params.numberOfElements,
        divisions: divisions,
      },
      points
    );
  }
  return points;
}

/*
params: {
    duration: <integer>,
    divisions: <integer>, 
    numberOfElements: <integer>,
    maxDuration: <float from 0 to 1>, // default: 0.5
    minDuration: <float from 0 to 1> // default: 0.1
}
return array,
0: duration
1: position
*/
export function timely(params) {
  // first we spread the starting point of our elements
  let startAndDuration = _addDimension(
    {
      numberOfElements: params.numberOfElements,
      from: (params.minDuration || minDurationDefault) * params.duration,
      to: (params.maxDuration || maxDurationDefault) * params.duration,
      divisions: params.divisions || defaultDivisions,
    },
    null
  );

  startAndDuration = _addDimension(
    {
      numberOfElements: params.numberOfElements,
      from: 0.3,
      to: 1,
      divisions: params.divisions || defaultDivisions,
    },
    startAndDuration
  );

  // now we are going to calculate the exact duration of each element
  // and replace the fraction that we have produced on the previous step

  // a variable to be used for making sure we make the last element's duration equal the number needed
  // so the full sequence has the desired duration
  let closestToEnd = {
    ms: 0,
    reference: null,
  };

  for (let i = 0; i < startAndDuration.length; i++) {
    const element = startAndDuration[i];
    element[0] = Math.round(element[0]);
    if (closestToEnd.ms < element[0]) {
      closestToEnd = {
        ms: element[0],
        reference: element,
      };
    }
    element[1] = Math.floor(
      params.duration - element[0] <
        params.duration * (params.maxDuration || maxDurationDefault)
        ? (params.duration - element[0]) * element[1]
        : (params.maxDuration || maxDurationDefault) *
            params.duration *
            element[1]
    );
  }
  closestToEnd.reference[1] = params.duration - closestToEnd.reference[0];

  return startAndDuration;
}


demo/index.js
import { HTMLClip, loadPlugin } from "@kissmybutton/motorcortex";
import Player from "@kissmybutton/motorcortex-player";
import MyPluginDefinition from "../dist/motorcortex-comets.umd";
const MyPlugin = loadPlugin(MyPluginDefinition);

const clip = new HTMLClip({
  html: `
    <div class="container">
      <div class="comet1"> </div>
      <div class="comet2"> </div>
    </div>`,
  css: `
    .container{
      width: 800px;
      height: 300px;
    }
  `,
  host: document.getElementById("clip"),
  containerParams: {
    width: "800px",
    height: "300px",
  },
});

const comet = new MyPlugin.Comets(
  {
    width: 800,
    height: 300,
    cometMaxSize: 1500,
    cometMinSize: 100,
    meteoriteColors: ["#FFFFFF", "#FFDC63"],
    auraIner: ["#57DDBA", "#62D8B9", "#7CC8B3"],
    auraOuter: ["#47F7C5", "#87B3C1"],
    items: 6,
    repeats: 1,
    duration: 8000,
  },
  {
    selector: ".comet1",
  }
);

const cometMagenta = new MyPlugin.Comets(
  {
    width: 800,
    height: 300,
    cometMaxSize: 1500,
    cometMinSize: 100,
    meteoriteColors: ["#FFFFFF", "#FFDC63"],
    auraIner: ["#AF4095", "#D48E93", "#CC9C9F"],
    auraOuter: ["#E06D95", "#E7B4A7"],
    items: 6,
    repeats: 1,
    duration: 8000,
  },
  {
    selector: ".comet2",
  }
);

const stars = new MyPlugin.Stars(
  {
    width: 800,
    height: 300,
    starMaxSize: 30,
    starMinSize: 5,
    starColor: "#F0F0F0",
    starGlowColor:
      "0 0 0 0 0.950946003   0 0 0 0 0.81265567   0 0 0 0 0.51528336  0 0 0 1 0",
    items: 20,
    repeats: 2,
    duration: 8000,
    blinkType: "opacity",
  },
  {
    selector: ".comet2",
  }
);

clip.addIncident(comet, 0);
clip.addIncident(cometMagenta, 0);
clip.addIncident(stars, 0);

new Player({ clip });
